projects/params_vs_compute/hash_ladder/hash_ladder.py

```diff
@@ -0,0 +1,477 @@
+
+parlai train_model -m projects.params_vs_compute.hash_ladder.hash_ladder:HashLadderAgent -t convai2:normalized -mf /tmp/model_file --ladder-size 1 --hash-size 32 --hash-layer 1
+"""
+
+
+class HashLadderAgent(TransformerGeneratorAgent):
+    """
+    Simple implementation of Hash Layers and the Ladder model from the following papers:
+    """
+
+    @classmethod
+    def add_cmdline_args(
+        cls, parser: ParlaiParser, partial_opt: Optional[Opt] = None
+    ) -> ParlaiParser:
+        TransformerGeneratorAgent.add_cmdline_args(parser, partial_opt=partial_opt)
+        # Add transformer args.
+        parser.add_argument(
+            '--ladder-size',
+            type=int,
+            default=1,
+            help='Number of ladder steps, default is not to use ladder.',
+        )
+        parser.add_argument(
+            '--hash-size', type=int, default=32, help='Number of hash bins.'
+        )
+        parser.add_argument(
+            '--hash-layer',
+            type=int,
+            default=7,
+            help='Layer number the Hash Layer appears on.',
+        )
+        return parser
+
+    def build_model(self, states=None):
+        wrapped_class = TransformerGeneratorModel.with_components(decoder=Decoder)
+        return wrapped_class(self.opt, self.dict)
+
+
+def _normalize(tensor, norm_layer):
+    """
+    Broadcast layer norm.
+    """
+    is_cpu = tensor.device == 'cpu' or tensor.device.type == 'cpu'
+    if APEX_LAYER_NORM and not is_cpu:
+        # fused_layer_norm has a bug around multi-device networks.
+        # https://github.com/NVIDIA/apex/issues/770
+        # https://github.com/NVIDIA/apex/issues/371
+        with torch.cuda.device(tensor.device):
+            return norm_layer(tensor)
+    else:
+        return norm_layer(tensor)
+
+
+class Decoder(TransformerDecoder):
+    """
+    Custom Decoder with Ladder model.
+    """
+
+    def __init__(
+        self,
+        opt: Opt,
+        embedding: Optional[nn.Embedding] = None,
+        n_positions: Optional[int] = None,
+        **kwargs,
+    ):
+        super().__init__(opt, **kwargs)
+
+        def _default(val, default):
+            return val if val is not None else default
+
+        opt['dict_size'] = embedding.weight.size(0)
+        self.opt = opt
+        self.embedding_size = opt['embedding_size']
+        self.ffn_size = opt['ffn_size']
+        self.n_layers = (
+            opt['n_decoder_layers']
+            if opt.get('n_decoder_layers', -1) > 0
+            else opt['n_layers']
+        )
+        self.n_heads = opt['n_heads']
+        self.dim = self.embedding_size
+        self.activation = opt.get('activation', 'relu')
+        self.variant = opt.get('variant', 'aiayn')
+
+        self.embeddings_scale = opt.get('embeddings_scale', True)
+        dropout_frac = opt.get('dropout', 0.0)
+        self.dropout = nn.Dropout(p=dropout_frac)  # --dropout
+
+        self.n_positions = _default(n_positions, get_n_positions_from_options(opt))
```

Comment:
Looks like many of these properties are the same as `TransformerDecoder` and you already call `super().init()`. Perhaps you only need the layer building code here.

I can add a `build_layers` function to `TransformerDecoder` in a subsequent PR for overriding purposes.

Reply:
ok, i'll look at these later after this initial merge, thanks